shared_buffer_pool.go

```diff
@@ -0,0 +1,153 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpc
+
+import "sync"
+
+// SharedBufferPool is a pool of buffers that can be shared, resulting in
+// decreased memory allocation. Currently, in gRPC-go, it is only utilized
+// for parsing incoming messages.
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type SharedBufferPool interface {
+	// Get returns a buffer with specified length from the pool.
+	Get(length int) []byte
+
+	// Put returns a buffer to the pool.
+	Put(*[]byte)
+}
+
+// NewSimpleSharedBufferPool creates a new SimpleSharedBufferPool with buckets
+// of different sizes to optimize memory usage. This prevents the pool from
+// wasting large amounts of memory, even when handling messages of varying sizes.
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+func NewSimpleSharedBufferPool() SharedBufferPool {
+	return &simpleSharedBufferPool{
+		pool0:   makeBytesPool(level0PoolMaxSize),
+		pool1:   makeBytesPool(level1PoolMaxSize),
+		pool2:   makeBytesPool(level2PoolMaxSize),
+		pool3:   makeBytesPool(level3PoolMaxSize),
+		pool4:   makeBytesPool(level4PoolMaxSize),
+		poolMax: makeFallbackBytesPool(),
+	}
+}
+
+// simpleSharedBufferPool is a simple implementation of SharedBufferPool.
+type simpleSharedBufferPool struct {
+	pool0   bufferPool
+	pool1   bufferPool
+	pool2   bufferPool
+	pool3   bufferPool
+	pool4   bufferPool
+	poolMax bufferPool
```

Comment:
How about a slice or array instead?  Along with a function to convert from size to bin, that would simplify Get and Put to 1/6th their current size.

Reply:
Done with 15f820e51f3622eb0a79185fd985d117f4f1c044